### PR TITLE
Leave card:

### PR DESCRIPTION
**Pull request by OleksandrGubariev**
1. add deleting row from cards_users table during deleting congratulations
2. add CongratulationDao to CardService
3. add method leaveCard to CardController
4. change datasets for links and types
5. add test for leaveCard controller
6. change test for method deleteByCardId in JdbcCongratulationDao

### CODE REVIEW

**Review by AlexKievDeveloper — APPROVED** [review]

**Review by RomanKlishch — APPROVED** [review]

File: src/main/java/com/greetingcard/web/controller/CardUserController.java [hunk 1]
```diff
@@ -42,4 +44,16 @@ public void deleteListMembers(@PathVariable long id, @RequestBody List<UserInfo>
         log.info("Users are successfully deleted from card {}", id);
     }
 
+    @DeleteMapping(value = "card/{id}/user")
+    public ResponseEntity<?> leaveCard(@PathVariable long id, HttpSession session) {
```
Comment (Yuliia-Sh) [inline src/main/java/com/greetingcard/web/controller/CardUserController.java:48]: "value=" is not needed here

**Review by AlexKievDeveloper — APPROVED** [review]

**Review by RomanKlishch — APPROVED** [review]

File: src/main/java/com/greetingcard/dao/jdbc/JdbcCongratulationDao.java [hunk 2]
```diff
@@ -9,34 +9,51 @@
...
+            "on cg.congratulation_id = links.congratulation_id WHERE card_id=?";
+    private static final String CHANGE_STATUS_CONGRATULATION_BY_CARD_ID = "UPDATE congratulations SET status_id = ? " +
+            "where card_id = ?";
+    private static final String CHANGE_CONGRATULATION_STATUS_BY_CONGRATULATION_ID = "UPDATE congratulations SET " +
+            "status_id = ? where congratulation_id = ?";
+    private static final String DELETE_BY_ID = "DELETE FROM congratulations WHERE " +
+            "congratulation_id= :congratulation_id and user_id= :user_id";
+    private static final String FIND_IMAGE_AND_AUDIO_LINKS_BY_CONGRATULATION_ID = "SELECT link FROM links l LEFT JOIN " +
+            "congratulations cg ON cg.congratulation_id = l.congratulation_id where cg.congratulation_id=? " +
+            "and (type_id = 2 OR type_id = 3) and user_id =?";
```
[the remark this answers is not in the document]
Comment (OleksandrGubariev) [reply src/main/java/com/greetingcard/dao/jdbc/JdbcCongratulationDao.java:52]: done

File: src/main/java/com/greetingcard/web/controller/CardUserController.java [hunk 3]
```diff
@@ -42,4 +44,16 @@ public void deleteListMembers(@PathVariable long id, @RequestBody List<UserInfo>
...
+        log.info("Request for leave card with id : {}", id);
+        User user = (User) session.getAttribute("user");
+
+        log.info("Request for leave card from user with id : {}", user.getId());
+        cardUserService.deleteUserFromCard(id, user.getId());
+
+        log.info("Successfully leave card with id: {}", id);
+        return ResponseEntity.status(HttpStatus.OK).build();
+    }
```
[the remark this answers is not in the document]
Comment (OleksandrGubariev) [reply src/main/java/com/greetingcard/web/controller/CardUserController.java:57]: done

File: src/test/java/com/greetingcard/web/controller/CardUserControllerSystemTest.java [hunk 4]
```diff
@@ -260,4 +260,15 @@ void deleteUsersEmptyListNotAdmin() throws Exception {
...
+    @DisplayName("Leave card")
+    void leaveCard() throws Exception {
+        User user = User.builder().id(1).build();
+        mockMvc.perform(delete("/api/v1/card/{id}/user", 1L)
+                .sessionAttr("user", user)
+                .characterEncoding("utf-8"))
+                .andDo(print())
+                .andExpect(status().isOk());
+    }
+
```
[the remark this answers is not in the document]
Comment (OleksandrGubariev) [reply src/test/java/com/greetingcard/web/controller/CardUserControllerSystemTest.java:273]: done

File: src/main/java/com/greetingcard/web/controller/CardUserController.java (hunk 1 above)
Comment (OleksandrGubariev) [reply src/main/java/com/greetingcard/web/controller/CardUserController.java:48]: done

File: src/main/java/com/greetingcard/dao/jdbc/JdbcCongratulationDao.java [hunk 5]
```diff
@@ -9,34 +9,51 @@
...
-    private static final String CHANGE_STATUS_CONGRATULATION_BY_CARD_ID = "UPDATE congratulations SET status_id = ? where card_id = ?";
-    private static final String CHANGE_CONGRATULATION_STATUS_BY_CONGRATULATION_ID = "UPDATE congratulations SET status_id = ? where congratulation_id = ?";
-    private static final String DELETE_BY_ID = "DELETE FROM congratulations WHERE congratulation_id=? and user_id=?";
-    private static final String FIND_IMAGE_AND_AUDIO_LINKS_BY_CONGRATULATION_ID = "SELECT link FROM links l LEFT JOIN congratulations cg ON cg.congratulation_id = l.congratulation_id where cg.congratulation_id=? and (type_id = 2 OR type_id = 3) and user_id =?";
+    private static final String LEAVE_BY_CARD_ID = "DELETE FROM congratulations WHERE card_id= :card_id and " +
+            "user_id= :user_id";
+    private static final String FIND_IMAGE_AND_AUDIO_LINKS_BY_CARD_ID = "SELECT link FROM links l LEFT JOIN " +
+            "congratulations cg ON cg.congratulation_id = l.congratulation_id where card_id=? " +
+            "and (type_id = 2 OR type_id = 3) and user_id =?";
```
[the remark this answers is not in the document]
Comment (OleksandrGubariev) [reply src/main/java/com/greetingcard/dao/jdbc/JdbcCongratulationDao.java:39]: done

**Review by nomarchia — APPROVED** [review]

**Pull request merged**